charms/sackd/src/charm.py

```diff
@@ -0,0 +1,163 @@
+            self._slurmctld.on.slurmctld_unavailable: self._on_slurmctld_unavailable,
+            self.on.service_sackd_started: self._on_sackd_started,
+            self.on.service_sackd_stopped: self._on_sackd_stopped,
+        }
+        for event, handler in event_handler_bindings.items():
+            self.framework.observe(event, handler)
+
+    def _on_install(self, event: InstallEvent) -> None:
+        """Perform installation operations for sackd."""
+        self.unit.status = WaitingStatus("installing sackd")
+
+        try:
+            self._sackd.install()
+            # Ensure sackd does not start before relation established
+            self._sackd.service.disable()
+            self.unit.set_workload_version(self._sackd.version())
+            self._systemd_notices.subscribe()
+            self._stored.sackd_installed = True
+        except SlurmOpsError as e:
+            logger.error(e.message)
+            event.defer()
+
+        self._check_status()
+
+    def _on_update_status(self, _: UpdateStatusEvent) -> None:
+        """Handle update status."""
+        self._check_status()
+
+    def _on_slurmctld_available(self, event: SlurmctldAvailableEvent) -> None:
+        """Retrieve the slurmctld_available event data and store in charm state."""
+        if self._stored.sackd_installed is not True:
+            event.defer()
+            return
+
+        if (slurmctld_host := event.slurmctld_host) != self._stored.slurmctld_host:
+            if slurmctld_host is not None:
+                self._sackd.config_server = f"{slurmctld_host}:6817"
+                self._stored.slurmctld_host = slurmctld_host
+                logger.debug(f"slurmctld_host={slurmctld_host}")
+            else:
+                logger.debug("'slurmctld_host' not in event data.")
+                return
+
+        if (auth_key := event.auth_key) != self._stored.auth_key:
+            if auth_key is not None:
+                self._stored.auth_key = auth_key
+                self._sackd.munge.key.set(auth_key)  # TODO change this once auth/slurm in place
+            else:
+                logger.debug("'auth_key' not in event data.")
+                return
+
+        logger.debug("#### Storing slurmctld_available event relation data in charm StoredState.")
+        self._stored.slurmctld_available = True
+
+        # Restart sackd after we write event data to respective locations.
+        self._sackd.munge.service.restart()  # TODO change this once auth/slurm in place
+        self._sackd.service.enable()
+        self._check_status()
+
+    def _on_slurmctld_unavailable(self, _) -> None:
+        """Stop sackd and set slurmctld_available = False when we lose slurmctld."""
+        logger.debug("## Slurmctld unavailable")
+        self._stored.slurmctld_available = False
+        self._stored.auth_key = ""
+        self._stored.slurmctld_host = ""
+        self._sackd.service.disable()
```

Comment:
I've not made this change since `self._stored.slurmctld_available = False` results in the `self._check_status()` call below setting `self.unit.status = WaitingStatus("Waiting on: slurmctld")`.

I think it makes sense not to set `BlockStatus` here and keep the `self._check_status()` call, especially as it's called every `on.update_status` so would replace the status we set here anyway. Another option would be to refactor `_check_status` to remove the `slurmctld_available` branch, or look to supplant the function in general, but that would require a bit of rethinking.

Reply:
Sure, we can keep it as is for now in `_check_status`. I think we'll need to figure out the whole holistic/"around the world" mechanism going forward as we'll likely need to expand it for when we start installing GPU and Infiniband drivers.